oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/authentication/OAuth2AccessTokenResponseAuthenticationSuccessHandler.java

```diff
@@ -0,0 +1,125 @@
+
+	private HttpMessageConverter<OAuth2AccessTokenResponse> accessTokenResponseConverter =
+			new OAuth2AccessTokenResponseHttpMessageConverter();
+
+	private Consumer<OAuth2AccessTokenAuthenticationContext> accessTokenResponseCustomizer;
+
+	@Override
+	public void onAuthenticationSuccess(HttpServletRequest request, HttpServletResponse response, Authentication authentication) throws IOException, ServletException {
+		if (!(authentication instanceof OAuth2AccessTokenAuthenticationToken accessTokenAuthentication)) {
+			if (this.logger.isErrorEnabled()) {
+				this.logger.error(Authentication.class.getSimpleName() + " must be of type " +
+						OAuth2AccessTokenAuthenticationToken.class.getName() +
+						" but was " + authentication.getClass().getName());
+			}
+			OAuth2Error error = new OAuth2Error(OAuth2ErrorCodes.SERVER_ERROR, "Unable to process the access token response.", null);
+			throw new OAuth2AuthenticationException(error);
+		}
+
+		OAuth2AccessToken accessToken = accessTokenAuthentication.getAccessToken();
+		OAuth2RefreshToken refreshToken = accessTokenAuthentication.getRefreshToken();
+		Map<String, Object> additionalParameters = accessTokenAuthentication.getAdditionalParameters();
+
+		OAuth2AccessTokenResponse.Builder builder =
+				OAuth2AccessTokenResponse.withToken(accessToken.getTokenValue())
+						.tokenType(accessToken.getTokenType())
+						.scopes(accessToken.getScopes());
+		if (accessToken.getIssuedAt() != null && accessToken.getExpiresAt() != null) {
+			builder.expiresIn(ChronoUnit.SECONDS.between(accessToken.getIssuedAt(), accessToken.getExpiresAt()));
+		}
+		if (refreshToken != null) {
+			builder.refreshToken(refreshToken.getTokenValue());
+		}
+		if (!CollectionUtils.isEmpty(additionalParameters)) {
+			builder.additionalParameters(additionalParameters);
+		}
+
+		if (this.accessTokenResponseCustomizer != null) {
+			// @formatter:off
+			OAuth2AccessTokenAuthenticationContext accessTokenAuthenticationContext =
+					OAuth2AccessTokenAuthenticationContext.with(accessTokenAuthentication)
+						.accessTokenResponse(builder)
+						.build();
+			// @formatter:on
+			this.accessTokenResponseCustomizer.accept(accessTokenAuthenticationContext);
+			if (this.logger.isTraceEnabled()) {
+				this.logger.trace("Customized access token response");
+			}
+		}
+
+		OAuth2AccessTokenResponse accessTokenResponse = builder.build();
+		ServletServerHttpResponse httpResponse = new ServletServerHttpResponse(response);
+		this.accessTokenResponseConverter.write(accessTokenResponse, null, httpResponse);
+	}
+
+	/**
+	 * Sets the {@link HttpMessageConverter} used for converting an {@link OAuth2AccessTokenResponse} to an HTTP response.
+	 *
+	 * @param accessTokenResponseConverter the {@link HttpMessageConverter} used for converting an {@link OAuth2AccessTokenResponse} to an HTTP response
+	 */
+	public void setAccessTokenResponseConverter(HttpMessageConverter<OAuth2AccessTokenResponse> accessTokenResponseConverter) {
```

Comment:
I'm for removing the Converter setter, however, I'm bumping into an issue as to how to properly test `onAuthenticationSuccess` method without being able to observe the Converter invocation (since there would be no way to override or inject a Converter, at least for testing purposes). Specifically [here](https://github.com/spring-projects/spring-authorization-server/pull/1429/files#diff-0ae510445a9f315af67faf0c1a8abc096ffe4f0c69528514e13316d1b55ef78cR112) and one other test.

Reply:
You can read the token response using this approach:

https://github.com/spring-projects/spring-authorization-server/blob/01441f917c131e0c91255fc0030703d5e6b81bbd/oauth2-authorization-server/src/test/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenEndpointFilterTests.java#L282

And then assert on the values in `OAuth2AccessTokenResponse`.

Reply:
Ah great, thanks for the heads up! I applied the change and it works nicely.